### DISCUSSION TITLE
ruser_mesh_data values on restart

### DISCUSSION THREAD

**Discussion by andreaantoni** [PrincetonUniversity/athena#470] [Q&A] [answered]
I have a time-dependent black hole interior to my spherical inner boundary (i.e. not on the grid) that is accreting mass through the inner boundary.  The black hole contributes to the gravity felt by the gas that is on the grid.   When I restart the simulation, I need to initialize the mass of the BH based on what it was at the time the restart file was generated.

Reading the Wiki, it seems like I should be able to store the BH mass in a ruser_mesh_data variable and that value will be written to the restart file and then available upon restart. Is that correct?

I am finding that after restart, the ruser_mesh_data variable is equal to 0 (it should be, e.g., 60) because the ruser_mesh_data arrays are allocated in InitUserMeshData, which is called again upon restart. 

Am I misunderstanding what is supposed to happen to these data upon restart?

Thanks!
Andrea

**Comment by tomidakn** [comment, marked as answer]
While the user-defined mesh/meshblock data are stored in a restart file and restored in the restarting process, 
> only the data on the master node (rank 0) is written in the restarting file, and it is users' responsibility to maintain consistency between nodes,

as written on the Wiki. So please make sure that you synchronize the data stored in ruser_mesh_data. You can do so, for example, using MPI_Allreduce in Mesh::UserWorkInLoop.

**Reply by andreaantoni** [reply]
Thank you.  Yes, I know that only the master node writes the data. 

I figured out the issue.  I was setting the final value of the mesh variable in UserWorkAfterLoop, but that value isn't what ends up in the restart file.   If I set it in UserWorkInLoop instead, that value *does* end up in the restart file. 

Thanks for your quick response!

**Reply by leeyp2009** [reply]
Could you share how you resolve this problem? I find that ruser_mesh_data is still the initial value when I update ruser_mesh_data in Mesh: UserWorkInLoop and synchronize the data with MPI_Allreduce. 

Upon restart, I suppose that I need to load the stored data from ruser_mesh_data in Mesh::InitUserMeshData, but I find it is still the initial value (e.g., zero) when I first initialize ruser_mesh_data before restart. 
The res_flag is not seen Mesh::InitUserMeshData, so I manually set a variable to determine when to read the stored data from ruser_mesh_data.  Many thanks!

**Reply by andreaantoni** [reply]
I don't do anything special upon restart.  Here is all I do with the ruser_mesh_data in my setup and I find that when I restart the simulation, ruser_mesh_data has the value that it was set to in the last call of Mesh::UserWorkInLoop before the prior simulation ended.  Note that I am using this in a slightly older version of Athena++ (the rad_newtonian branch from before it was merged with master a ~month ago).

1) allocate and initialize at the start of the simulation (cycle=0; this isn't called again on restart):
```
void  Mesh::InitUserMeshData(ParameterInput *pin) {
  AllocateRealUserMeshDataField(1);
  ruser_mesh_data[0].NewAthenaArray(1);  
  ruser_mesh_data[0](0) = some_value;
}
```

2) While the simulation is running, I update the value on CPUs and all meshblocks to the same value: 
```
void Mesh::UserWorkInLoop() {

 #ifdef MPI_PARALLEL
    Real tempdm;
    MPI_Allreduce(&dm_global, &tempdm, 1, MPI_ATHENA_REAL, MPI_SUM, MPI_COMM_WORLD);
    dm_global = tempdm;
#endif

Real &m_base_global = ruser_mesh_data[0](0);
m_base_global += dm_global;
dm_global=0;

return;
}
```

This doesn't seem very different from what you are doing, so maybe there is something subtle.
Have you tried doing a test on 1 MeshBlock and 1 CPU?

**Reply by leeyp2009** [reply]
Hi Andrea, 
Many thanks for your quick response! I have tested with a similar approach as yours. But the Mesh::InitUserMeshData is actually called again for restart runs. So the ruser_mesh_data is re-set to the initial value upon restart (e.g., the "some_value" you mentioned above) in Mesh::UserWorkInLoop. So I am wondering how could you restore the m_base_global data before the restart? Thanks!

**Reply by leeyp2009** [reply]
The problem is resolved. There is indeed something subtle here. I forget to save the ruser_mesh_data in some temporary array before updating them. Thanks, Andrea.

**Reply by andreaantoni** [reply]
Glad it works for you now!  

For others who may be wondering why the variables are correct on restart even though `Mesh::InitUserMeshData` is called on restart (which is where you would set the initial values at the very beginning of the simulation:

In src/mesh/mesh.cpp, the mesh generator for restarts, with function prototype `Mesh::Mesh(ParameterInput *pin, IOWrapper& resfile, int mesh_test)`, does call `InitUserMeshData(pin);` but then it immediately overwrites the, e.g., ruser_mesh_data arrays with the values from the restart file, e.g. lines ~756-760:
```
    for (int n=0; n<nreal_user_mesh_data_; n++) {
      std::memcpy(ruser_mesh_data[n].data(), &(userdata[udoffset]),
                  ruser_mesh_data[n].GetSizeInBytes());
      udoffset += ruser_mesh_data[n].GetSizeInBytes();
    }
    delete [] userdata;
  }
```

**Comment by kyriakosTapinou** [comment]
@tomidakn 

I don't understand the wiki comment 
"only the data on the master node (rank 0) is written in the restarting file, and it is users' responsibility to maintain consistency between nodes,"

Why have ruser_meshblock_data for each meshblock and then not save each on uniquely?

I am trying to restart where I need each meshblock's ruser data from the restart file loaded back into memory before restarting.

**Reply by tomidakn** [reply]
Sorry, the statement was not accurate - it applies only to user_mesh_data, and user_meshblock_data are correctly stored (and automatically restored) in a restart file.

**Reply by kyriakosTapinou** [reply]
@tomidakn thank you for your reply. 

I'm currently refactoring my data-driving fork of Athena++ and I've been avoiding the restart infrastructure for a while - I'm digging through it now to ensure it functions correctly. 

To your knowledge is there anything in the baseline Athena++ that would "get angry" from specifying varying sizes of ruser_meshblocks across different blocks?

**Reply by tomidakn** [reply]
Unfortunately, it will get angry, because the current implementation assumes that all MeshBlocks have the same data size.

To be more specific, it uses MPI IO and it needs to know the size and position of each MeshBlock data. To simplify this process I assumed they are all same size. You need to change the file structure to store the size (and position) of each MeshBlock, or add padding to Make all MeshBlocks same size. I think the latter is redundant but simpler.

**Reply by msbc** [reply]
I accidentally did this before because I only needed user mesh block data on the edge of the simulation domain. I either got IO or seg-fault errors when I tried. I just ended up allocating this data on all mesh blocks and just ignoring the user meshblock data that was on most of the meshblocks.

**Reply by kyriakosTapinou** [reply]
Ok I see, I for now I've just been copying to all blocks as well @msbc but the productions are allocating GB of data through ruser_meshblocks_data so I can't ignore this inefficiency forever in my personal implementation hahaha. 

@tomidakn I see - this is something I'll have to look into eventually - I will post once I have it working.
